python/pyspark/ml/tests/connect/test_parity_torch_distributor.py

```diff
@@ -0,0 +1,511 @@
+        inputs = [
+            ("0,1,2", 2, True, "0"),
+        ]
+
+        for i, (_, num_processes, use_gpu, expected) in enumerate(inputs):
+            with self.subTest(f"subtest: {i + 1}"):
+                dist = TorchDistributor(num_processes, False, use_gpu, spark=self.spark)
+                dist._run_training_on_pytorch_file = lambda *args: os.environ.get(
+                    CUDA_VISIBLE_DEVICES, "NONE"
+                )
+                self.assertEqual(
+                    expected,
+                    dist._run_distributed_training(dist._run_training_on_pytorch_file, "..."),
+                )
+
+    # Can not set SparkConf
+    # def test_get_num_tasks_distributed(self) -> None:
+    #     inputs = [(1, 8, 8), (2, 8, 4), (3, 8, 3)]
+    #
+    #     for spark_conf_value, num_processes, expected_output in inputs:
+    #         with self.subTest():
+    #             self.spark.sparkContext._conf.set(
+    #                 "spark.task.resource.gpu.amount", str(spark_conf_value)
+    #             )
+    #             distributor = TorchDistributor(num_processes, False, True, self.spark)
+    #             self.assertEqual(distributor._get_num_tasks(), expected_output)
+    #
+    #     self.spark.sparkContext._conf.set("spark.task.resource.gpu.amount", "1")
+
+    def test_distributed_file_with_pytorch(self) -> None:
+        test_file_path = "python/test_support/test_pytorch_training_file.py"
+        learning_rate_str = "0.01"
+        TorchDistributor(num_processes=2, local_mode=False, use_gpu=False, spark=self.spark).run(
+            test_file_path, learning_rate_str
+        )
+
+    def test_end_to_end_run_distributedly(self) -> None:
+        train_fn = create_training_function(self.mnist_dir_path)
+        output = TorchDistributor(
+            num_processes=2, local_mode=False, use_gpu=False, spark=self.spark
+        ).run(train_fn, 0.001)
+        self.assertEqual(output, "success")
+
+
+@unittest.skipIf(not have_torch, "torch is required")
+class TorchWrapperUnitTests(unittest.TestCase):
+    def test_clean_and_terminate(self) -> None:
+        def kill_task(task: "subprocess.Popen") -> None:
+            time.sleep(1)
+            clean_and_terminate(task)
+
+        command = [sys.executable, "-c", '"import time; time.sleep(20)"']
+        task = subprocess.Popen(command)
+        t = threading.Thread(target=kill_task, args=(task,))
+        t.start()
+        time.sleep(2)
+        self.assertEqual(task.poll(), 0)  # implies task ended
+
+    @patch("pyspark.ml.torch.torch_run_process_wrapper.clean_and_terminate")
+    def test_check_parent_alive(self, mock_clean_and_terminate: Callable) -> None:
+        command = [sys.executable, "-c", '"import time; time.sleep(2)"']
+        task = subprocess.Popen(command)
+        t = threading.Thread(target=check_parent_alive, args=(task,), daemon=True)
+        t.start()
+        time.sleep(2)
+        self.assertEqual(mock_clean_and_terminate.call_count, 0)
```

Comment:
Q:

Why we need to copy the testing code ? Can we reuse original test code like DataFrame connect test suite ?

Reply:
we can reuse it for sure, right now I just copy it for convinience.
when this pr is ready, it will reuse existing ut